src/Compilers/Core/Portable/Serialization/StreamObjectReader.cs

```diff
@@ -0,0 +1,1020 @@
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(kind);
+            }
+        }
+
+        private class VariantListReader : ObjectReader
+        {
+            private readonly List<Variant> _list;
+            private int _index;
+
+            public VariantListReader()
+            {
+                _list = new List<Variant>();
+            }
+
+            public List<Variant> List => _list;
+
+            public void Reset()
+            {
+                _list.Clear();
+                _index = 0;
+            }
+
+            public int Position => _index;
+
+            private Variant Next()
+            {
+                return _list[_index++];
+            }
+
+            public override bool ReadBoolean()
+            {
+                return Next().AsBoolean();
+            }
+
+            public override byte ReadByte()
+            {
+                return Next().AsByte();
+            }
+
+            public override char ReadChar()
+            {
+                return Next().AsChar();
+            }
+
+            public override decimal ReadDecimal()
+            {
+                return Next().AsDecimal();
+            }
+
+            public override double ReadDouble()
+            {
+                return Next().AsDouble();
+            }
+
+            public override float ReadSingle()
+            {
+                return Next().AsSingle();
+            }
+
+            public override int ReadInt32()
+            {
+                return Next().AsInt32();
+            }
+
+            public override long ReadInt64()
+            {
+                return Next().AsInt64();
+            }
+
+            public override sbyte ReadSByte()
+            {
+                return Next().AsSByte();
+            }
+
+            public override short ReadInt16()
+            {
+                return Next().AsInt16();
+            }
+
+            public override uint ReadUInt32()
+            {
+                return Next().AsUInt32();
+            }
+
+            public override ulong ReadUInt64()
+            {
+                return Next().AsUInt64();
+            }
+
+            public override ushort ReadUInt16()
+            {
+                return Next().AsUInt16();
+            }
+
+            public override DateTime ReadDateTime()
+            {
+                return Next().AsDateTime();
+            }
+
+            public override String ReadString()
+            {
+                var next = Next();
+                if (next.Kind == VariantKind.Null)
+                {
+                    return null;
+                }
+                else
+                {
+                    return next.AsString();
+                }
+            }
+
+            public override Object ReadValue()
+            {
+                return Next().ToBoxedObject();
+            }
+        }
+
+        private class ReferenceMap : IDisposable
```

Comment:
This seems super complex vs just using a dictionary.  is there a reason we need this complexity?